lib/src/simple_op_store.rs

```diff
@@ -125,6 +119,18 @@ impl OpStore for SimpleOpStore {
     }
 }
 
+pub fn read_view_from_file(view_path: PathBuf) -> OpStoreResult<View> {
+    let buf = fs::read(view_path)?;
+
+    let proto = crate::protos::op_store::View::decode(&*buf)?;
+    Ok(view_from_proto(proto))
+}
+
+pub fn write_view_to_file(view: &View, path: PathBuf) -> OpStoreResult<()> {
+    let proto = view_to_proto(view);
+    Ok(std::fs::write(path, proto.encode_to_vec())?)
+}
+
```

Comment:
I'm not sure about this approach. The way it is now, `git.rs` already has to deal with `crate::op_store::View`, and `view::View`. Following your suggestion would add `crate::protos::op_store::View` to the list. I'd rather go the other way.

I'm currently thinking about creating a special error type just for this. I'd probably put it in `view.rs`, together with shims over the two functions I have here. Then, `git.rs` can only deal with `view::View` and functions form `view.rs`.

WDYT? Is this too many shims?

Reply:
I had thought the point of these functions was to reduce duplication, but I see about the various `View` types. I agree that that's confusing.

Actually, maybe we should define a separate proto for what we need here instead of reusing the one from `simple_op_store`? Since this models the last seen Git state, am I right in thinking that we need map from ref name to commit id, and maybe a git HEAD target too?

Reply:
We could have a map from git refs to commit ids. In the future, we might want to track not only branches but also tags and remote-tracking branches. I guess these are all git refs.

The main question is whether we'd want to track anything about jj's side of the state in this file (e.g. if this is conflicted). My guess is no. Another way to ask this question: can you think of any part of jj's state that shouldn't be affected by `jj undo`? I don't think I can.

This would be a significant change to this PR, but probably it's easier to do now rather than later. Do I need a crash course on creating a new proto or is it pretty straightforward?

Reply:
Also, isn't git's HEAD also a ref? I don't know if its official ref name is just "HEAD". Or is it special?

Reply:
> Also, isn't git's HEAD also a ref?

I think it's not a regular but a "symbolic ref". I *think* all real refs have to start with `refs/`.


> This would be a significant change to this PR, but probably it's easier to do now rather than later.

Yeah, sorry I didn't think of it earlier :(

> Do I need a crash course on creating a new proto or is it pretty straightforward to use prost? I'll look into it.

It's pretty straightforward. Just add a proto in `lib/src/protos/`, update the list `lib/gen-protos/src/main.rs`, and then  go into `lib/gen-protos/` and run `cargo run`. That should create a new `.rs` file in `lib/src/`. I think that's it.

> The main question is whether we'd want to track anything about jj's side of the state in this file

No, since the file is there to track what we've seen in the git repo, so we don't overwrite state on import or export, I think it should just be the git state that we need to record.

Reply:
I guess there could be more than one HEAD if there are several git worktrees. Not sure how we'd want to track it; ideally it'd be as similar to what Git does as possible but I don't know what that is. Or, we could track it in the same dictionary but start it with something other than refs/ until we figure out how to do it properly (e.g. `jj-synthetic-ref/HEAD`) and whether we want to bother with Git worktrees.

Reply:
Good point. We [don't support that yet](https://github.com/martinvonz/jj/blob/5351371d51c7f83727c944b7d0bf4e08118fe2ec/lib/src/op_store.rs#L200), and protos are pretty flexible (as I'm sure you know), so we should be able to replace a single `bytes git_head = 2` field by `repeated GitHead git_heads = 3` field (or similar) if we add support for that.

Reply:
We hereby decree that every git repo shall have at most one head. For now.

Makes sense, given that other protos already follow this decree.

Reply:
I have a draft in https://github.com/ilyagr/jj/tree/protos. The last commit has some confusing test failures (see it's description). They may be due to subtle changes in behavior since the last seen git refs are no longer allowed to be conflicted.

It might take me a bit before I can get to carefully debugging them.

Reply:
Ok, now the PR comes with its own proto. It wasn't too bad.